### ISSUE TITLE
vagrant-berkshelf should only be enabled if Berksfile is present

### ISSUE THREAD

**Issue by tknerr** [berkshelf/vagrant-berkshelf#261]
Just wondering, wasn't it the default behaviour that vagrant-berkshelf was only enabled when a `Berksfile` is present?

I'm noticing that vagrant-berkshelf *always* kicks in, sets up a shelf, and modifies the chef provisioner path, even if there is no Berksfile present at all!

This is vagrant-berkshelf 4.0.2 with vagrant 1.7.2

Is this the expected behaviour?

**Comment by sethvargo** [comment]
@tknerr the logic for [enabling automatically is here](https://github.com/berkshelf/vagrant-berkshelf/blob/master/lib/vagrant-berkshelf/config.rb#L51-L67). The behavior you described is the expected behavior. We also have unit tests for this, so if you are experiencing alternate behavior, it's likely that Vagrant does not load things the way I suspect. Are you seeing it actually _run_ the `berks` command?

**Comment by tknerr** [comment]
@sethvargo nope, it does not run the `berks` command, it just sets up the shelf and modifies the chef provisioner's `cookbooks_path` to point to the shelf.

Minimal sample Vagrantfile, no Berksfile present in the same directory:
```ruby
Vagrant.configure(2) do |config|
  config.vm.box = "chef/ubuntu-12.04-i386"
  config.vm.provision "chef_solo" do |chef|
    chef.cookbooks_path = "./cookbooks"
  end
end
```

So I get this (notice the shelf being set up but no cookbook vendoring) :
```
C:\Repos\_stash\vagrant-chef-excercises\10-chef-solo-with-vagrant\solution\tmp>vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
    default: The Berkshelf shelf is at "W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-15592-88s2a7-default"
==> default: Sharing cookbooks with VM
==> default: Importing base box 'chef/ubuntu-12.04-i386'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: tmp_default_1422890571040_81670
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 => 2222 (adapter 1)
...
```

In contrast, when a Berksfile is present, I see it actually run the `berks vendor` command, e.g. here:
```
C:\Repos\_github\dev-box>vagrant up
Bringing machine 'dev-box' up with 'virtualbox' provider...
    dev-box: The Berkshelf shelf is at "W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box"
==> dev-box: Sharing cookbooks with VM
==> dev-box: Importing base box 'tknerr/ubuntu1404-desktop'...
==> dev-box: Matching MAC address for NAT networking...
==> dev-box: Checking if box 'tknerr/ubuntu1404-desktop' is up to date...
==> dev-box: Setting the name of the VM: dev-box_dev-box_1422879971270_14297
==> dev-box: Updating Vagrant's Berkshelf...
==> dev-box: Resolving cookbook dependencies...
==> dev-box: Fetching 'dev-box' from source at .
==> dev-box: Using apt (1.3.2)
==> dev-box: Using bash (0.1.2)
==> dev-box: Using chef-dk (3.0.0)
==> dev-box: Using chef_handler (1.1.6)
==> dev-box: Using dmg (2.2.2)
==> dev-box: Using dev-box (0.1.0) from source at .
==> dev-box: Using vagrant (0.2.0)
==> dev-box: Using windows (1.36.1)
==> dev-box: Vendoring apt (1.3.2) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/apt
==> dev-box: Vendoring bash (0.1.2) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/bash
==> dev-box: Vendoring chef-dk (3.0.0) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/chef-dk
==> dev-box: Vendoring chef_handler (1.1.6) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/chef_handler
==> dev-box: Vendoring dev-box (0.1.0) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/dev-box
==> dev-box: Vendoring dmg (2.2.2) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/dmg
==> dev-box: Vendoring vagrant (0.2.0) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/vagrant
==> dev-box: Vendoring windows (1.36.1) to W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-7240-r7zhsc-dev-box/windows
==> dev-box: Clearing any previously set network interfaces...
==> dev-box: Preparing network interfaces based on configuration...
    dev-box: Adapter 1: nat
==> dev-box: Forwarding ports...
    dev-box: 22 => 2222 (adapter 1)
...
```

**Comment by tknerr** [comment]
Added some printline debugging around the code you mentioned, and it looks the shelf is being set up even before the config is validated:
```
C:\Repos\_stash\vagrant-chef-excercises\10-chef-solo-with-vagrant\solution\tmp>vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
    default: The Berkshelf shelf is at "W:/home/.berkshelf/vagrant-berkshelf/shelves/berkshelf20150202-15592-88s2a7-default"
==> default: Sharing cookbooks with VM
@enabled is #<Object:0x404ab10>
@berksfile_path is
path is C:/Repos/_stash/vagrant-chef-excercises/10-chef-solo-with-vagrant/solution/tmp/Berksfile
==> default: Importing base box 'chef/ubuntu-12.04-i386'...
==> default: Matching MAC address for NAT networking...
@enabled is false
@berksfile_path is
==> default: Setting the name of the VM: tmp_default_1422890571040_81670
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 => 2222 (adapter 1)
...
```

**Comment by tknerr** [comment]
Guess I found the culprit here:
https://github.com/berkshelf/vagrant-berkshelf/blob/master/lib/vagrant-berkshelf/helpers.rb#L121

It looks like `helpers.berkshelf_enabled?` is called before the config is validated. At this point in time `env[:machine].config.berkshelf.enabled` is `MAYBE`, which is truthy and thus `berkshelf_enabled?` is.

I changed it to this and it looks better now (still testing, though):
```ruby
      def berkshelf_enabled?(env)
         env[:machine].config.berkshelf.enabled == true
      end
```

**Comment by sethvargo** [comment]
Ah - yea - we should explicitly check that since `Object.new` is never nil.

**Comment by tknerr** [comment]
this would probably fix #254 as well
